text/0069-api-v2.md

```diff
@@ -0,0 +1,178 @@
+# TiKV API V2
+
+## Motivation
+
+`API V2` is a set of breaking changes that aim to solve serval issues with current RawKV (hereafter referred to as `API V1`):
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Document-level design
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+### Upgrade
+
+Upgrade from `API V1` to `API V2` is a simple process:
+
+1. Update TiKV, TiDB, and PD to the version that supports `API V2`.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any.
+3. Ensure that no non-TiDB client is online.
+4. Use `pd-ctl` to enable `API V2`.
+5. Enable `API V2` in TiKV config file and restart TiKV.
+
+### Downgrade
+
+Downgrade from `API V2` to `API V1` is also simple:
+
+1. Ensure that no non-TiDB TiKV client is online.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any.
+3. Use `pd-ctl` to disable `API V2`.
+4. Disable `API V2` in TiKV config file and restart TiKV.
+
+### CDC / BR
+
+Since all access to TiDB is unchanged during the upgrade, CDC and BR should work trivally on TiDB data.
+
+### Data migration
+
+It's reasonable to provide a way to import and export non-TiDB data in TiKV during the upgrade or downgrade. On TiKV before 4.0, the only way to do that is `scan` and `batch_put` on the client. After 4.0, TiKV start to support importing SST file into TxnKV, and after 5.1, importing on RawKV is also supported. You can find more information in [`RFC: Online Bulk Load for RawKV`](https://github.com/tikv/rfcs/pull/72).
```

Comment:
> How to distinguish and scan all the non-tidb data in TiKV? If there is non-tidb data inside a tikv cluster 4.0

The TxnKV data start with `m` or `t` are considered as tidb data, otherwise are non-tidb data.

>  If there is non-tidb data inside a tikv cluster 4.0, what's the correct step to update the cluster to the latest version without data losing?

The user should migrate the non-tidb to new version tikv either by the new `SST import and export` feature or `get()` and `put()` through the client.

Reply:
if users want to upgrade a tikv cluster to a newer version. should they first scan all the existing non-tidb data from tikv, store them to external storage (maybe a file), then upgrade the cluster, and finally put these non-tidb data from the external storage to the upgraded tikv cluster?

Reply:
This is one of the approches.

Reply:
you can put other reasonable approaches to this section as well.